### ISSUE TITLE
Make EmitC generated module headers match normal header-only library style.

### ISSUE THREAD

**Issue by benvanik** [google/iree#7605]
The common way to do header-only libraries is to have a .h that contains the declarations for the public interface as well as the definitions but with the definitions guarded such that they are only included by the user on-demand (this is how the compiler tablegen files work, too). Today the emitc headers are really like .c files and cannot be included by the user in more than one file. This also creates badness like #7603 where the users will need to include the module header in whatever compilation unit they want to use it in which may not be compatible with the generated code.

Concretely this would mean adding an include guard, guarding the entirety of the current .h output (including #includes of the internal vm headers like vm_ops.h) with an `#if defined(IMPLEMENTATION)`, and adding a declaration of the module create function that is non-static:
```c
#ifndef GENERATED_SOURCE_FILE_NAME_H_
#define GENERATED_SOURCE_FILE_NAME_H_

#include "iree/vm/api.h"

#ifdef __cplusplus
extern "C" {
#endif  // __cplusplus

// this is the only function users need, and is safe to include in a .cc because
// of the extern c and lack of transitive includes of libc goo
iree_status_t add_module_create(iree_allocator_t v1, iree_vm_module_t** v2);

#ifdef __cplusplus
}  // extern "C"
#endif  // __cplusplus

#endif  // GENERATED_SOURCE_FILE_NAME_H_

#if defined(GENERATED_SOURCE_FILE_NAME_IMPL)
#include "iree/vm/ops.h"
// the entire rest of the existing file
#endif  // GENERATED_SOURCE_FILE_NAME_IMPL
```

Then a user can include the header anywhere they want and just create the module, like in `my_test.cc`:
```c
#include "generated_source.h"
...
  IREE_RETURN_IF_ERROR(add_module_create(...));
```

And then where they want the implementation - in a .c or a safe .cc:
```c
#define GENERATED_SOURCE_FILE_NAME_IMPL
#include "generated_source.h"
```

The easiest thing to do would be add the prefix/suffix as a postprocessing step after emitc produces the body of the file so that we don't have to teach emitc to do all of this. The prefix/suffix will be uniform besides the templated names.

**Comment by marbre** [comment]
Thanks for the detailed issue description @benvanik!

Applying your suggestion to [`iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp`](https://github.com/google/iree/blob/main/iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp) would be straight forward and probably that is what I'll do in a first step. However, `CModuleTarget.cpp` needs some more care anyway. For example we should get rid of https://github.com/google/iree/blob/f6e9188af3d341a47a8d223ec2031c996dd742d1/iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp#L355-L358 and replace those with `emitc::IncludeOp`s (which get printed by the Cpp emitter). If we want to do this + apply your suggestion, we probably need teach emitc to do (parts) of what you suggested. Even this would be fine for me, since we should improve where we see the need!.

**Comment by marbre** [comment]
Fixed with #7929.
